crates/server/src/lib.rs

```diff
@@ -120,60 +122,67 @@ impl Config {
 /// Represents the warg registry server.
 pub struct Server {
     config: Config,
-    listener: Option<TcpListener>,
+    endpoints: Option<Endpoints>,
+    token: CancellationToken,
+    tasks: JoinSet<Result<()>>,
+}
+
+/// The bound endpoints for the warg registry server.
+#[derive(Clone)]
+pub struct Endpoints {
+    pub api: SocketAddr,
 }
 
 impl Server {
     /// Creates a new server with the given configuration.
     pub fn new(config: Config) -> Self {
         Self {
             config,
-            listener: None,
+            token: CancellationToken::new(),
+            endpoints: None,
+            tasks: JoinSet::new(),
         }
     }
 
-    /// Binds the server to the configured address.
+    /// Starts the server and binds its endpoint to the configured address.
     ///
-    /// Returns the address the server bound to.
-    pub fn bind(&mut self) -> Result<SocketAddr> {
+    /// Returns the endpoints the server bound to.
+    pub async fn start(&mut self) -> Result<Endpoints> {
+        assert!(
+            self.endpoints.is_none(),
+            "cannot start server multiple times"
+        );
```

Comment:
So I'm personally not a big fan of taking `&mut self` for this as it leads to confusing internal state and needing to wrap things in `Option` just for this method to take their contents.

What if we renamed this `run` method to `spawn` and in addition to returning the endpoints, we return a type wrapping both the `JoinSet` representing the spawned server's tasks and the `CancellationToken` that can be used to stop the server. Thus, the type could have a `stop` method on it to stop the server and also implement `Future` that a caller can await on to join the tasks in the set?

Reply:
Agree, I was just about to ask why required fields were becoming optional. I'd much rather have spawning produce a dedicated type that represents the info for a started service.

Reply:
The existing code for `bind` method is `&mut self`, but the `run` method was `self`, both are now combined into your `start` method.

That said, I think perhaps we could better model our `Server` off of hyper's [Server](https://docs.rs/hyper/latest/hyper/server/struct.Server.html); theirs is _already_ a bounds and listening server upon construction thus there doesn't need to be a `start` (or `spawn`) method.

That is, you contruct/build a server, it internally binds based on its configuration; the binding can be inspected via a method; in the case of hype there is only a [single bound address](https://docs.rs/hyper/latest/hyper/server/struct.Server.html#method.local_addr) for the server, but our implementation might have multiple bound addresses as there might be multiple hyper servers internally (e.g. API, monitoring, etc).

Finally, it implements `Future`, so that you simply await it for it to run to graceful completion.